### PR TITLE
Merge20220812

### PR DESCRIPTION
**Pull request by samtertzakian**
1. Work around issue where WDF sometimes calls QuerRemove callback twice.
2. Fix some asserts.
3. Add DMF_Time Module. More Methods to be added later.
4. During RemoveComplete close WDFIOTARGET before calling Client callback so view is the same as if QueryRemove had happened.

### CODE REVIEW

File: Dmf/Modules.Library/Dmf_Time.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
+## DMF_Stack
```
Comment (ankurung) [inline Dmf/Modules.Library/Dmf_Time.md:1]: Time.

File: Dmf/Modules.Library/Dmf_Time.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+-----------------------------------------------------------------------------------------------------------------------------------
+
+Implements a time keeper module which gets current cpu ticks and also calculates the time elapsed in milliseconds and nanoseconds.
+
+-----------------------------------------------------------------------------------------------------------------------------------
+
+#### Module Configuration
+
+-----------------------------------------------------------------------------------------------------------------------------------
+##### DMF_CONFIG_Stack
```
Comment (ankurung) [inline Dmf/Modules.Library/Dmf_Time.md:16]: Time

File: Dmf/Modules.Library/Dmf_Time.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+````
+typedef struct
+{
+    // Keeping empty for now. Might add something later.
+    //
+} DMF_CONFIG_Stack;
```
Comment (ankurung) [inline Dmf/Modules.Library/Dmf_Time.md:22]: Time

**Review by ankurung — COMMENTED** [review]
DMF_Time documentation needs an update..

File: Dmf/Modules.Library/Dmf_Time.c [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+                         SECONDS_TO_NANOSECONDS,
+                         &result);
+    if (hResult != S_OK)
+    {
+        DmfAssert(false);
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "Integer overflow when performing multiplication of %lld and %lld", elaspedTicks, SECONDS_TO_NANOSECONDS);
+        ntStatus = STATUS_INTEGER_OVERFLOW;
+        goto Exit;
+    }
+#endif
```
Comment (ankurung) [inline Dmf/Modules.Library/Dmf_Time.c:126]: where are we making status success for UserMode?

File: Dmf/Modules.Library/Dmf_Time.md (hunk 1 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_Time.md:1]: fixed

File: Dmf/Modules.Library/Dmf_Time.md (hunk 2 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_Time.md:16]: removed

File: Dmf/Modules.Library/Dmf_Time.c (hunk 4 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_Time.c:126]: fixed

**Review by ankurung — APPROVED** [review]

**Pull request merged**